homeassistant/components/binary_sensor/tod.py

```diff
@@ -0,0 +1,250 @@
+"""
+Support for showing the binary sensors represending current time of the day.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/binary_sensor.tod/
+
+"""
+from datetime import datetime, timedelta
+import logging
+import pytz
+
+import voluptuous as vol
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDevice, ENTITY_ID_FORMAT, PLATFORM_SCHEMA)
+from homeassistant.const import (
+    CONF_FRIENDLY_NAME, CONF_AFTER, CONF_BEFORE,
+    CONF_SENSORS, SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET)
+from homeassistant.core import callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import async_generate_entity_id
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.helpers.sun import (
+    get_astral_event_date, get_astral_event_next)
+from homeassistant.util import dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_AFTER_OFFSET = 'after_offset'
+CONF_BEFORE_OFFSET = 'before_offset'
+
+ATTR_NEXT_UPDATE = 'next_update'
+
+
+SENSOR_SCHEMA = vol.Schema({
+    vol.Required(CONF_AFTER): vol.Any(cv.time, vol.All(
+        vol.Lower, cv.sun_event)),
+    vol.Optional(CONF_AFTER_OFFSET, default=timedelta(0)): cv.time_period,
+    vol.Required(CONF_BEFORE): vol.Any(cv.time, vol.All(
+        vol.Lower, cv.sun_event)),
+    vol.Optional(CONF_BEFORE_OFFSET, default=timedelta(0)): cv.time_period,
+    vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SENSORS): vol.Schema({cv.slug: SENSOR_SCHEMA}),
+})
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the ToD sensors."""
+    if hass.config.time_zone is None:
+        _LOGGER.error("Timezone is not set in Home Assistant configuration")
+        return
+
+    sensors = []
+    for name, device_config in config[CONF_SENSORS].items():
+        after = device_config[CONF_AFTER]
+        after_offset = device_config[CONF_AFTER_OFFSET]
+        before = device_config[CONF_BEFORE]
+        before_offset = device_config[CONF_BEFORE_OFFSET]
+        friendly_name = device_config.get(CONF_FRIENDLY_NAME, name)
+        sensors.append(
+            TodSensor(
+                hass, name, friendly_name,
+                after, after_offset, before, before_offset
+            )
+        )
+
+    async_add_entities(sensors)
+
+
+def is_sun_event(event):
+    """Return true if event is sun event not time."""
+    return event in (SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET)
+
+
+class TodSensor(BinarySensorDevice):
+    """Time of the Day Sensor."""
+
+    def __init__(self, hass, name, friendly_name, after, after_offset,
+                 before, before_offset):
+        """Init the ToD Sensor..."""
+        self.hass = hass
```

Comment:
Rather than passing in hass, please override `async_added_to_hass` and set up the initial values there.

Reply:
Moved the value initialisation to async_added_to_hass. Removed passing in hass.